[Parallel-Safe] Improve JSON parser robustness for LLM output

## Scope (json_parser.py only)

Improve JSON parsing to handle common LLM output errors. This is **safe for parallel work** as it only touches the utility file.

## File to modify

- `src/utils/json_parser.py`

## Issues to address

### 1. List-to-dict conversion
LLM sometimes returns a list instead of expected dict object.

**Evidence:**
```
[ERROR] Item refinement returned invalid JSON structure: ['Attunement required...', 'Requires encryption...']
```

**Fix:** Add fallback that wraps list in a dict with appropriate key:
```python
if isinstance(parsed, list):
    # Try to infer the expected structure
    return {"properties": parsed}  # or similar contextual wrapping
```

### 2. Better truncation recovery
Current truncation detection works but could be improved.

**Evidence:**
```
[WARNING] Detected truncated JSON (unbalanced braces) - likely token limit reached
```

**Fix:** 
- Add more aggressive brace/bracket closing
- Try multiple repair strategies in sequence
- Log which strategy succeeded for debugging

### 3. Add structured error information
Return error details alongside partial results:
```python
@dataclass
class ParseResult:
    data: dict | list | None
    success: bool
    repair_applied: str | None  # e.g., "truncation_repair", "list_to_dict"
    original_error: str | None
```

## Testing

Add tests in `tests/unit/test_utils/test_json_parser.py`:
- Test list-to-dict conversion
- Test various truncation scenarios
- Test nested object recovery

## Labels

- `enhancement`
- `copilot-safe`

## Related

- Part of #153 (JSON parsing failures)

--- a/src/utils/json_parser.py
+++ b/src/utils/json_parser.py
@@ -32,20 +32,49 @@ def convert_list_to_dict(
 
     # If list contains dicts, it might be a list of entities - return first
     if all(isinstance(item, dict) for item in data):
-        logger.warning(f"List contains {len(data)} dict items, returning first as entity")
-        return data[0] if data else {}
+        context_info = f" for context '{context_hint}'" if context_hint else ""
+        logger.warning(
+            f"List contains {len(data)} dict items, returning first as entity{context_info}"
+        )
+        first_item: dict[str, Any] = data[0]
+        return first_item
 
     # If list contains strings, treat as properties/features
     if all(isinstance(item, str) for item in data):
-        logger.warning(f"Converting list of {len(data)} strings to properties dict")
+        context_info = f" for context '{context_hint}'" if context_hint else ""
+        logger.warning(f"Converting list of {len(data)} strings to properties dict{context_info}")
         return {
             "properties": data,
             "description": "; ".join(str(x) for x in data[:3]) + ("..." if len(data) > 3 else ""),
         }
 
     # Mixed types - convert to generic structure
-    logger.warning(f"Converting mixed-type list of {len(data)} items to dict")
+    context_info = f" for context '{context_hint}'" if context_hint else ""
+    logger.warning(f"Converting mixed-type list of {len(data)} items to dict{context_info}")
     return {"items": data}
+
+
+def _count_unescaped_quotes(text: str) -> int:
+    """Count unescaped double quotes in text.
+
+    Args:
+        text: String to analyze.
+
+    Returns:
+        Number of unescaped quote characters.
+    """
+    count = 0
+    escape_next = False
+    for char in text:
+        if escape_next:
+            escape_next = False
+            continue
+        if char == "\\":
+            escape_next = True
+            continue
+        if char == '"':
+            count += 1
+    return count
 
 
 def _repair_truncated_json(text: str) -> dict[str, Any] | list[Any] | None:
@@ -66,8 +95,8 @@ def _repair_truncated_json(text: str) -> dict[str, Any] | list[Any] | None:
 
         if open_braces > 0 or open_brackets > 0:
             repaired = text.rstrip(",: \n\t\"'")
-            # Close any open string
-            if repaired.count('"') % 2 == 1:
+            # Close any open string (escape-aware)
+            if _count_unescaped_quotes(repaired) % 2 == 1:
                 repaired += '"'
             repaired += "]" * max(0, open_brackets) + "}" * max(0, open_braces)
             result = _try_parse_json(repaired)
@@ -94,7 +123,7 @@ def _repair_truncated_json(text: str) -> dict[str, Any] | list[Any] | None:
             if char == "\\":
                 escape_next = True
                 continue
-            if char == '"' and not escape_next:
+            if char == '"':
                 in_string = not in_string
                 continue
             if in_string:
@@ -358,17 +387,22 @@ def parse_json_to_model[T](
     if data is None:
         return None
 
-    if not isinstance(data, dict):
-        error_msg = f"Expected JSON object but got {type(data).__name__}"
-        if strict:
-            logger.error(error_msg)
-            raise JSONParseError(
-                error_msg,
-                response_preview=response[:500],
-                expected_type=model_class.__name__,
-            )
-        logger.debug(error_msg)
-        return None
+    if isinstance(data, list):
+        # LLM returned list instead of dict - try to convert
+        logger.warning(
+            f"Expected JSON object but got list for {model_class.__name__} - attempting conversion"
+        )
+        data = convert_list_to_dict(data, context_hint=model_class.__name__)
+        if not data:
+            error_msg = f"Expected JSON object but got empty list for {model_class.__name__}"
+            if strict:
+                logger.error(error_msg)
+                raise JSONParseError(
+                    error_msg,
+                    response_preview=response[:500],
+                    expected_type=model_class.__name__,
+                )
+            return None
 
     try:
         return model_class(**data)

--- a/tests/unit/test_json_parser.py
+++ b/tests/unit/test_json_parser.py
@@ -252,19 +252,47 @@ class TestParseJsonToModel:
         result = parse_json_to_model(response, SampleModel, strict=False)
         assert result is None
 
-    def test_raises_error_when_json_is_list_strict(self):
-        """Should raise JSONParseError when JSON is a list, not object, in strict mode."""
+    def test_converts_list_to_model_when_first_item_matches(self):
+        """Should convert list to dict and create model when first item has all fields."""
+        response = """```json
+[{"name": "first", "value": 1}, {"name": "second", "value": 2}]
+```"""
+        result = parse_json_to_model(response, SampleModel)
+        assert isinstance(result, SampleModel)
+        assert result.name == "first"
+        assert result.value == 1
+
+    def test_raises_error_when_converted_list_fails_model_strict(self):
+        """Should raise JSONParseError when converted list doesn't match model in strict mode."""
         response = """```json
 [{"name": "a"}, {"name": "b"}]
 ```"""
         with pytest.raises(JSONParseError) as exc_info:
             parse_json_to_model(response, SampleModel)
-        assert "Expected JSON object but got list" in str(exc_info.value)
+        # The converted dict is {"name": "a"} which lacks "value" field
+        assert "Failed to create SampleModel" in str(exc_info.value)
 
-    def test_returns_none_when_json_is_list_non_strict(self):
-        """Should return None when JSON is a list, not object, in non-strict mode."""
+    def test_returns_none_when_converted_list_fails_model_non_strict(self):
+        """Should return None when converted list doesn't match model in non-strict mode."""
         response = """```json
 [{"name": "a"}, {"name": "b"}]
+```"""
+        result = parse_json_to_model(response, SampleModel, strict=False)
+        assert result is None
+
+    def test_raises_error_when_list_is_empty_strict(self):
+        """Should raise JSONParseError when JSON is an empty list in strict mode."""
+        response = """```json
+[]
+```"""
+        with pytest.raises(JSONParseError) as exc_info:
+            parse_json_to_model(response, SampleModel)
+        assert "empty list" in str(exc_info.value)
+
+    def test_returns_none_when_list_is_empty_non_strict(self):
+        """Should return None when JSON is an empty list in non-strict mode."""
+        response = """```json
+[]
 ```"""
         result = parse_json_to_model(response, SampleModel, strict=False)
         assert result is None

